// @observablehq/notebook-kit/dist/src/runtime/stdlib/recommendedLibraries.js
export const _ = () => import("npm:lodash").then((_) => _.default);
export const aq = () => import("npm:arquero");
export const Arrow = () => import("npm:apache-arrow");
export const d3 = () => import("npm:d3");
export const dot = () => import("./dot.js").then((_) => _.dot);
export const duckdb = () => import("npm:@duckdb/duckdb-wasm");
export const DuckDBClient = () => import("./duckdb.js").then((_) => _.DuckDBClient);
export const echarts = () => import("npm:echarts");
export const htl = () => import("npm:htl");
export const html = () => import("npm:htl").then((_) => _.html);
export const svg = () => import("npm:htl").then((_) => _.svg);
export const Inputs = () => import("./inputs.js");
export const L = () => import("./leaflet.js");
export const mapboxgl = () => import("./mapboxgl.js").then((_) => _.default);
export const md = () => import("./md.js").then((_) => _.md);
export const mermaid = () => import("./mermaid.js").then((_) => _.mermaid);
export const Plot = () => import("npm:@observablehq/plot");
export const React = () => import("npm:react");
export const ReactDOM = () => import("npm:react-dom");
// export const sql = () => import("observablehq:stdlib/duckdb").then((_) => _.sql);
// export const SQLite = () => import("observablehq:stdlib/sqlite").then((_) => _.default);
// export const SQLiteDatabaseClient = () => import("observablehq:stdlib/sqlite").then((_) => _.SQLiteDatabaseClient);
export const tex = () => import("./tex.js").then((_) => _.tex);
export const topojson = () => import("npm:topojson-client");
// export const vg = () => import("observablehq:stdlib/vgplot").then((_) => _.default());
export const vl = () => import("./vega-lite.js").then((_) => _.vl);
